lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -1264,11 +1264,93 @@ class MuxPad : public mlir::RewritePattern {
     return success();
   }
 };
+
+// Find muxes which have conditions dominated by other muxes with the same
+// condition.
+class MuxSharedCond : public mlir::RewritePattern {
+public:
+  MuxSharedCond(MLIRContext *context)
+      : RewritePattern(MuxPrimOp::getOperationName(), 0, context) {}
+
+  Value updateOrClone(MuxPrimOp mux, Value high, Value low,
+                      mlir::PatternRewriter &rewriter, bool chain) const {
+    if (chain) {
+      rewriter.updateRootInPlace(mux, [&] {
+        mux.setOperand(1, high);
+        mux.setOperand(2, low);
+      });
+      return {};
+    }
+    rewriter.setInsertionPointAfter(mux);
+    return rewriter
+        .create<MuxPrimOp>(mux.getLoc(), mux.getType(),
+                           ValueRange{mux.getSel(), high, low})
+        .getResult();
+  }
+
+  // Walk a dependent mux tree assuming the condition cond is true.
+  Value tryCondTrue(Value op, Value cond, mlir::PatternRewriter &rewriter,
+                    bool chain) const {
```

Comment:
There is a lot of duplication in the `tryCondTrue` and `tryCondFalse` methods. Is there a way to clean it up? I started at this for a while and couldn't come up with anything obvious. This is not blocking.